x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1933,4 +1954,154 @@ private static void typeResolutions(
+                return AggregateMetricDoubleBlockBuilder.Metric.MIN;
+            }
+            if (expression instanceof Sum || expression instanceof SumOverTime) {
+                return AggregateMetricDoubleBlockBuilder.Metric.SUM;
+            }
+            if (expression instanceof Count || expression instanceof CountOverTime || expression instanceof AvgOverTime) {
+                return AggregateMetricDoubleBlockBuilder.Metric.COUNT;
+            }
+            return null;
+        }
+
+        private Expression resolveMetricFunction(Expression expression, AggregateMetricDoubleBlockBuilder.Metric metric) {
+            AggregateFunction aggregateFunction = (AggregateFunction) expression;
+            if (aggregateFunction.field() instanceof FieldAttribute fa
+                && fa.field() instanceof InvalidMappedField imf
+                && typesShouldBeConverted(imf.types())) {
+                HashMap<ResolveUnionTypes.TypeResolutionKey, Expression> typeResolutions = new HashMap<>();
+                typeResolutions(imf, aggregateFunction, fa, expression, metric, typeResolutions);
+                var resolvedField = ResolveUnionTypes.resolvedMultiTypeEsField(fa, typeResolutions);
+                var newFieldAttribute = createIfDoesNotAlreadyExist(fa, resolvedField, unionFieldAttributes, metric);
+                return expression.replaceChildren(List.of(newFieldAttribute, expression.children().get(1)));
+            }
+            return expression;
+        }
+
+        private Expression substituteSurrogates(Expression expression) {
+            AggregateFunction aggregateFunction = (AggregateFunction) expression;
+            if (aggregateFunction.field() instanceof FieldAttribute fa && fa.field() instanceof InvalidMappedField imf) {
+                if (typesShouldBeConverted(imf.types()) == false) {
+                    return expression;
+                }
+                return SubstituteSurrogateExpressions.rule(expression);
+            }
+            return expression;
+        }
+
+        private boolean typesShouldBeConverted(Set<DataType> types) {
+            if (types.contains(AGGREGATE_METRIC_DOUBLE) == false) {
+                return false;
+            }
+            for (DataType type : types) {
+                if (type.isNumeric() == false && type != AGGREGATE_METRIC_DOUBLE) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private void typeResolutions(
+            InvalidMappedField imf,
+            AggregateFunction aggregateFunction,
+            FieldAttribute fa,
+            Expression expression,
+            AggregateMetricDoubleBlockBuilder.Metric metric,
+            HashMap<ResolveUnionTypes.TypeResolutionKey, Expression> typeResolutions
+        ) {
+            for (DataType type : imf.types()) {
+                // Effectively the contents of ResolveUnionTypes::typeSpecificConvert(...)
```

Comment:
I think we can fix `ResolveUnionTypes::typeSpecificConvert` (the number of children is not always 1), make FromAggregateMetricDouble a ConvertFunction, and avoid this fork.